
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="Genomics Aotearoa">
      
      
        <link rel="canonical" href="https://genomicsaotearoa.github.io/Introduction-to-R/appendix/04-bioconductor-vcfr/">
      
      
        <link rel="prev" href="../../08-r-help/">
      
      
        <link rel="next" href="../05-dplyr/">
      
      
      <link rel="icon" href="../../theme_figures/nesi_ga.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.16">
    
    
      
        <title>Using packages from Bioconductor - Introduction to R</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.7e37652d.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Mukta:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Mukta";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
   <link href="../../assets/stylesheets/glightbox.min.css" rel="stylesheet"/><style>
    html.glightbox-open { overflow: initial; height: 100%; }
    .gslide-title { margin-top: 0px; user-select: text; }
    .gslide-desc { color: #666; user-select: text; }
    .gslide-image img { background: white; }
    .gscrollbar-fixer { padding-right: 15px; }
    .gdesc-inner { font-size: 0.75rem; }
    body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color);}
    body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color);}
    body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color);}</style> <script src="../../assets/javascripts/glightbox.min.js"></script></head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#using-packages-from-bioconductor" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Introduction to R" class="md-header__button md-logo" aria-label="Introduction to R" data-md-component="logo">
      
  <img src="../../theme_figures/nesi_ga.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Introduction to R
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Using packages from Bioconductor
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m14.3 16-.7-2h-3.2l-.7 2H7.8L11 7h2l3.2 9zM20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12zm-9.15 3.96h2.3L12 9z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/GenomicsAotearoa/Introduction-to-R" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.0.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path fill="currentColor" d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/Introduction-to-R
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Introduction to R" class="md-nav__button md-logo" aria-label="Introduction to R" data-md-component="logo">
      
  <img src="../../theme_figures/nesi_ga.png" alt="logo">

    </a>
    Introduction to R
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/GenomicsAotearoa/Introduction-to-R" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.0.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path fill="currentColor" d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/Introduction-to-R
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../00-introduction/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introducing R and RStudio IDE
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../01-r-basics/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    R Basics
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../02-data-prelude/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to the example dataset and file type
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../03-basics-factors-dataframes/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    R Basics continued - factors and data frames
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../06-data-visualization/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Data Visualization with ggplot2
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../07-knitr-markdown/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Producing Reports With knitr
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../08-r-help/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Getting help with R
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_9" checked>
        
          
          <label class="md-nav__link" for="__nav_9" id="__nav_9_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Appendix
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_9_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_9">
            <span class="md-nav__icon md-icon"></span>
            Appendix
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Using packages from Bioconductor
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Using packages from Bioconductor
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#installing-packages-from-somewhere-else-besides-cran" class="md-nav__link">
    <span class="md-ellipsis">
      Installing packages from somewhere else besides CRAN?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#so-how-do-i-use-it" class="md-nav__link">
    <span class="md-ellipsis">
      So how do I use it?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#first-install-the-biocmanager-package" class="md-nav__link">
    <span class="md-ellipsis">
      First, install the BiocManager package
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#second-install-the-vcfr-package-from-bioconductor-using-biocmanager" class="md-nav__link">
    <span class="md-ellipsis">
      Second, install the vcfR package from Bioconductor using BiocManager
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#search-for-bioconductor-packages-based-on-your-analysis-needs" class="md-nav__link">
    <span class="md-ellipsis">
      Search for Bioconductor packages based on your analysis needs
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#resources" class="md-nav__link">
    <span class="md-ellipsis">
      Resources
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../05-dplyr/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Data Wrangling and Analyses with Tidyverse
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#installing-packages-from-somewhere-else-besides-cran" class="md-nav__link">
    <span class="md-ellipsis">
      Installing packages from somewhere else besides CRAN?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#so-how-do-i-use-it" class="md-nav__link">
    <span class="md-ellipsis">
      So how do I use it?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#first-install-the-biocmanager-package" class="md-nav__link">
    <span class="md-ellipsis">
      First, install the BiocManager package
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#second-install-the-vcfr-package-from-bioconductor-using-biocmanager" class="md-nav__link">
    <span class="md-ellipsis">
      Second, install the vcfR package from Bioconductor using BiocManager
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#search-for-bioconductor-packages-based-on-your-analysis-needs" class="md-nav__link">
    <span class="md-ellipsis">
      Search for Bioconductor packages based on your analysis needs
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#resources" class="md-nav__link">
    <span class="md-ellipsis">
      Resources
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="using-packages-from-bioconductor">Using packages from Bioconductor<a class="headerlink" href="#using-packages-from-bioconductor" title="Permanent link">&para;</a></h1>
<div class="admonition info">
<p class="admonition-title">Info</p>
<div class="tabbed-set tabbed-alternate" data-tabs="1:3"><input checked="checked" id="__tabbed_1_1" name="__tabbed_1" type="radio" /><input id="__tabbed_1_2" name="__tabbed_1" type="radio" /><input id="__tabbed_1_3" name="__tabbed_1" type="radio" /><div class="tabbed-labels"><label for="__tabbed_1_1">Keyp oints</label><label for="__tabbed_1_2">Objectives</label><label for="__tabbed_1_3">Questions</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<ul>
<li>Bioconductor is an alternative package repository for bioinformatics packages.</li>
<li>Installing packages from Bioconductor requires a new method, since it is not compatible with the <code>install.packages()</code> function used for CRAN.</li>
<li>Check Bioconductor to see if there is a package relevant to your analysis before writing code yourself.</li>
</ul>
</div>
<div class="tabbed-block">
<ul>
<li>Describe what the Bioconductor repository is and what it is used for</li>
<li>Describe how Bioconductor differs from CRAN</li>
<li>Search Bioconductor for relevant packages</li>
<li>Install a package from Bioconductor</li>
</ul>
</div>
<div class="tabbed-block">
<ul>
<li>How do I use packages from the Bioconductor repository?</li>
</ul>
</div>
</div>
</div>
</div>
<h2 id="installing-packages-from-somewhere-else-besides-cran">Installing packages from somewhere else besides CRAN?<a class="headerlink" href="#installing-packages-from-somewhere-else-besides-cran" title="Permanent link">&para;</a></h2>
<p>So far we have told you about using packages that are included in the
base installation of R (this is what comes with R 'out of the box'), and
packages that you can install from <a href="https://cran.r-project.org/">CRAN</a>
(the Comprehensive R Archive Network), which is the primary place many
people look for supplemental R packages to install. However, not all R
packages are available on CRAN. For bioinformatics-related packages in
particular, there is another repository that has many powerful packages
that you can install. It is called
<a href="https://bioconductor.org/">Bioconductor</a> and it is a repository
specifically focused on bioinformatics packages.
<a href="https://bioconductor.org/">Bioconductor</a> has a mission of
"promot[ing] the statistical analysis and comprehension of current and
emerging high-throughput biological assays." This means that many if not
all of the packages available on Bioconductor are focused on the
analysis of biological data, and that it can be a great place to look
for tools to help you analyze your -omics datasets!</p>
<h2 id="so-how-do-i-use-it">So how do I use it?<a class="headerlink" href="#so-how-do-i-use-it" title="Permanent link">&para;</a></h2>
<p>Since access to the <a href="https://bioconductor.org/">Bioconductor</a> repository
is not built in to base R 'out of the box', there are a couple steps
needed to install packages from this alternative source. We will work
through the steps (only 2!) to install a package to help with the VCF
analysis we are working on, but you can use the same approach to install
any of the many thousands of available packages.</p>
<p><a class="glightbox" href="../../figures/bioconductor_website_screenshot.jpg" data-type="image" data-width="auto" data-height="auto" data-desc-position="bottom"><img alt="bioconductor_homepage" src="../../figures/bioconductor_website_screenshot.jpg" /></a></p>
<h2 id="first-install-the-biocmanager-package">First, install the <code>BiocManager</code> package<a class="headerlink" href="#first-install-the-biocmanager-package" title="Permanent link">&para;</a></h2>
<p>The first step is to install a package that <em>is</em> on CRAN, <code>BiocManager</code>.
This package will allow us to use it to install packages from
Bioconductor. You can think of Bioconductor kind of like an alternative
app store for your phone, except instead of apps you are installing
packages, and instead of your phone it's your local R package library.</p>
<div class="admonition r-project">
<p class="admonition-title">r</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="c1"># Install the BiocManager from CRAN using the base R install.packages() function </span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a><span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">)</span>
</code></pre></div>
</div>
<p>To check if this worked (and also so you can make a note of the version
for reproducibility purposes), you can run <code>BiocManager::version()</code> and
it should give you the version number.</p>
<h2 id="second-install-the-vcfr-package-from-bioconductor-using-biocmanager">Second, install the vcfR package from Bioconductor using <code>BiocManager</code><a class="headerlink" href="#second-install-the-vcfr-package-from-bioconductor-using-biocmanager" title="Permanent link">&para;</a></h2>
<div class="admonition tip">
<p class="admonition-title">Head's Up: Installing vcfR may take a while due to numerous dependencies</p>
<p>Just be aware that installing packages that have many dependencies can take a while.</p>
</div>
<div class="admonition r-project">
<p class="admonition-title">r</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a><span class="c1"># install the vcfR package from Bioconductor </span>
<a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a><span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;vcfR&quot;</span><span class="p">)</span>
</code></pre></div>
</div>
<p>Depending on your particular system, you may need to also allow it to
install some dependencies or update installed packages in order to
successfully complete the process.</p>
<div class="admonition note">
<p class="admonition-title">Installing packages from Bioconductor vs from CRAN</p>
<p>Some packages begin by being available only on Bioconductor, and then later move to CRAN. <code>vcfR</code> is one such package, which originally was only available from Bioconductor, but is currently available from CRAN. The other thing to know is that <code>BiocManager::install()</code> will also install packages from CRAN (it is a wrapper around <code>install.packages()</code> that adds some extra features). There are other benefits to using <code>BiocManager::install()</code> for Bioconductor packages, many of which are outlined <a href="https://www.bioconductor.org/install/">here</a>. In short, Bioconductor packages have a release cycle that is different from CRAN and the <code>install()</code> function is aware of that difference, so it helps to keep package versions in line with one another in a way that doesn't generally happen with the base R <code>install.packages()</code>.</p>
</div>
<h2 id="search-for-bioconductor-packages-based-on-your-analysis-needs">Search for Bioconductor packages based on your analysis needs<a class="headerlink" href="#search-for-bioconductor-packages-based-on-your-analysis-needs" title="Permanent link">&para;</a></h2>
<p>While we are only focusing in this workshop on VCF analyses, there are
hundreds or thousands of different types of data and analyses that
bioinformaticians may want to work with. Sometimes you may get a new
dataset and not know exactly where to start with analyzing or
visualizing it. The Bioconductor package search view can be a great way
to browse through the packages that are available.</p>
<p><a class="glightbox" href="../../figures/bioconductor_search.jpg" data-type="image" data-width="auto" data-height="auto" data-desc-position="bottom"><img alt="bioconductor_search" src="../../figures/bioconductor_search.jpg" /></a></p>
<div class="admonition tip">
<p class="admonition-title">Searching for packages on the Bioconductor website</p>
<p>There are several thousand packages available through the Bioconductor website. It can be a bit of a challenge to find what you want, but one helpful resource is the <a href="https://bioconductor.org/packages/release/BiocViews.html#___Software">package search page</a>.</p>
</div>
<p>In bioinformatics, there are often many different tools that can be used
in a particular instance. The authors of <code>vcfR</code> have <a href="https://github.com/knausb/vcfR#software-that-produce-vcf-files">compiled some of
them</a>.
One of those packages that <a href="https://bioconductor.org/packages/release/bioc/html/VariantAnnotation.html">is available from
Bioconductor</a>
is called <code>VariantAnnotation</code> and may also be of interest to those
working with vcf files in R.</p>
<div class="admonition question">
<p class="admonition-title">Challenge</p>
<ul>
<li>Use the <code>BiocManager::available()</code> function to see what packages are available matching a search term.</li>
<li>Use the <a href="https://bioconductor.org/packages/release/BiocViews.html#___Software">biocViews</a> interface to search for packages of interest.</li>
</ul>
<p>You may or may not want to try installing the package, since not all dependencies always install easily. However, this will at least let you see what is available.</p>
</div>
<div class="admonition tip">
<p class="admonition-title">Refreshing the RStudio package view after installing</p>
<p>If you install a package from Bioconductor, you may need to refresh the RStudio package view to see it in your list. You can do this by clicking the <kbd>Refresh</kbd> button in <kbd>Packages</kbd> pane of RStudio.</p>
</div>
<h2 id="resources">Resources<a class="headerlink" href="#resources" title="Permanent link">&para;</a></h2>
<ul>
<li><a href="https://bioconductor.org/">Bioconductor</a></li>
<li><a href="https://bioconductor.org/packages/release/BiocViews.html#___Software">Bioconductor package
    search</a></li>
<li><a href="https://cran.r-project.org/">CRAN</a></li>
</ul>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tracking", "navigation.sections", "navigation.expand", "navigation.top"], "search": "../../assets/javascripts/workers/search.d50fe291.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.50899def.min.js"></script>
      
        <script src="../../javascripts/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  <script id="init-glightbox">const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});
document$.subscribe(() => { lightbox.reload() });
</script></body>
</html>