
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="Genomics Aotearoa">
      
      
        <link rel="canonical" href="https://genomicsaotearoa.github.io/Introduction-to-R/02-data-prelude/">
      
      
        <link rel="prev" href="../01-r-basics/">
      
      
        <link rel="next" href="../03-basics-factors-dataframes/">
      
      
      <link rel="icon" href="../theme_figures/nesi_ga.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.16">
    
    
      
        <title>Introduction to the example dataset and file type - Introduction to R</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.7e37652d.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Mukta:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Mukta";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../stylesheets/extra.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
   <link href="../assets/stylesheets/glightbox.min.css" rel="stylesheet"/><style>
    html.glightbox-open { overflow: initial; height: 100%; }
    .gslide-title { margin-top: 0px; user-select: text; }
    .gslide-desc { color: #666; user-select: text; }
    .gslide-image img { background: white; }
    .gscrollbar-fixer { padding-right: 15px; }
    .gdesc-inner { font-size: 0.75rem; }
    body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color);}
    body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color);}
    body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color);}</style> <script src="../assets/javascripts/glightbox.min.js"></script></head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#introduction-to-the-example-dataset-and-file-type" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Introduction to R" class="md-header__button md-logo" aria-label="Introduction to R" data-md-component="logo">
      
  <img src="../theme_figures/nesi_ga.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Introduction to R
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Introduction to the example dataset and file type
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m14.3 16-.7-2h-3.2l-.7 2H7.8L11 7h2l3.2 9zM20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12zm-9.15 3.96h2.3L12 9z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/GenomicsAotearoa/Introduction-to-R" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.0.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path fill="currentColor" d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/Introduction-to-R
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Introduction to R" class="md-nav__button md-logo" aria-label="Introduction to R" data-md-component="logo">
      
  <img src="../theme_figures/nesi_ga.png" alt="logo">

    </a>
    Introduction to R
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/GenomicsAotearoa/Introduction-to-R" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.0.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path fill="currentColor" d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/Introduction-to-R
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../00-introduction/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introducing R and RStudio IDE
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../01-r-basics/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    R Basics
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Introduction to the example dataset and file type
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Introduction to the example dataset and file type
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#preface" class="md-nav__link">
    <span class="md-ellipsis">
      Preface
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#introduction-to-the-dataset" class="md-nav__link">
    <span class="md-ellipsis">
      Introduction to the dataset
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-vcf-files-are-generated" class="md-nav__link">
    <span class="md-ellipsis">
      How VCF files are generated
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-variant-calls-are-stored-in-vcf-files" class="md-nav__link">
    <span class="md-ellipsis">
      How variant calls are stored in VCF files
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../03-basics-factors-dataframes/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    R Basics continued - factors and data frames
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../06-data-visualization/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Data Visualization with ggplot2
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../07-knitr-markdown/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Producing Reports With knitr
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../08-r-help/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Getting help with R
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_9" >
        
          
          <label class="md-nav__link" for="__nav_9" id="__nav_9_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Appendix
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_9">
            <span class="md-nav__icon md-icon"></span>
            Appendix
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../appendix/04-bioconductor-vcfr/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Using packages from Bioconductor
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../appendix/05-dplyr/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Data Wrangling and Analyses with Tidyverse
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#preface" class="md-nav__link">
    <span class="md-ellipsis">
      Preface
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#introduction-to-the-dataset" class="md-nav__link">
    <span class="md-ellipsis">
      Introduction to the dataset
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-vcf-files-are-generated" class="md-nav__link">
    <span class="md-ellipsis">
      How VCF files are generated
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-variant-calls-are-stored-in-vcf-files" class="md-nav__link">
    <span class="md-ellipsis">
      How variant calls are stored in VCF files
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="introduction-to-the-example-dataset-and-file-type">Introduction to the example dataset and file type<a class="headerlink" href="#introduction-to-the-example-dataset-and-file-type" title="Permanent link">&para;</a></h1>
<div class="admonition info">
<p class="admonition-title">Learning outcomes</p>
<div class="tabbed-set tabbed-alternate" data-tabs="1:3"><input checked="checked" id="__tabbed_1_1" name="__tabbed_1" type="radio" /><input id="__tabbed_1_2" name="__tabbed_1" type="radio" /><input id="__tabbed_1_3" name="__tabbed_1" type="radio" /><div class="tabbed-labels"><label for="__tabbed_1_1">Key points</label><label for="__tabbed_1_2">Objectives</label><label for="__tabbed_1_3">Questions</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<ul>
<li>The dataset comes from a real world experiment in <em>E. coli</em>.</li>
<li>Publicly available FASTQ files can be downloaded from NCBI SRA.</li>
<li>Several steps are taken outside of R/RStudio to create VCF files from
  FASTQ files.</li>
<li>VCF files store variant calls in a special format.</li>
</ul>
</div>
<div class="tabbed-block">
<ul>
<li>Know what the example dataset represents</li>
<li>Know the concepts of how VCF files are generated</li>
</ul>
</div>
<div class="tabbed-block">
<ul>
<li>What data are we using in the lesson?</li>
<li>What are VCF files?</li>
</ul>
</div>
</div>
</div>
</div>
<h2 id="preface">Preface<a class="headerlink" href="#preface" title="Permanent link">&para;</a></h2>
<p>The Intro to R and RStudio for Genomics is a part of the Genomics Data
Carpentry lessons. In this lesson we will learn the necessary skill sets
for R and RStudio and apply them directly to a real next-generation
sequencing (NGS) data in the variant calling format (VCF) file type.
Previous Genomics Data Carpentry lessons teach learners how to generate
a VCF file from FASTQ files downloaded from NCBI Sequence Read Archive
(SRA), so we won't cover that here. Instead, in this episode we will
give a brief overview of the data and a what VCF file types are for
those who wish to teach the Intro to R and RStudio for Genomics lesson
independently of the Genomics Data Carpentry lessons.</p>
<p>This dataset was selected for several reasons, including:</p>
<ul>
<li>Simple, but iconic NGS-problem: Examine a population where we want
    to characterize changes in sequence <em>a priori</em></li>
<li>Dataset publicly available - in this case through the NCBI SRA
    (<a href="http://www.ncbi.nlm.nih.gov/sra">http://www.ncbi.nlm.nih.gov/sra</a>)</li>
</ul>
<h2 id="introduction-to-the-dataset">Introduction to the dataset<a class="headerlink" href="#introduction-to-the-dataset" title="Permanent link">&para;</a></h2>
<p>Microbes are ideal organisms for exploring 'Long-term Evolution
Experiments' (LTEEs) - thousands of generations can be generated and
stored in a way that would be virtually impossible for more complex
eukaryotic systems. In <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4988878/">Tenaillon et al.
(2016)</a>, 12
populations of <em>Escherichia coli</em> were propagated for more than 50,000
generations in a glucose-limited minimal medium. This medium was
supplemented with citrate which <em>E. coli</em> cannot metabolize in the
aerobic conditions of the experiment. Sequencing of the populations at
regular time points reveals that spontaneous citrate-using mutants
(Cit+) appeared in a population of <em>E.coli</em> (designated Ara-3) at around
31,000 generations. It should be noted that spontaneous Cit+ mutants are
extraordinarily rare - inability to metabolize citrate is one of the
defining characters of the <em>E. coli</em> species. Eventually, Cit+ mutants
became the dominant population as the experimental growth medium
contained a high concentration of citrate relative to glucose. Around
the same time that this mutation emerged, another phenotype become
prominent in the Ara-3 population. Many <em>E. coli</em> began to develop
excessive numbers of mutations, meaning they became hypermutable.</p>
<p>Strains from generation 0 to generation 50,000 were sequenced, including
ones that were both Cit+ and Cit- and hypermutable in later generations.</p>
<p>For the purposes of this workshop we're going to be working with 3 of
the sequence reads from this experiment.</p>
<table>
<thead>
<tr>
<th>SRA run number</th>
<th>Clone</th>
<th>Generation</th>
<th>Cit</th>
<th>Hypermutable</th>
<th>Read length</th>
<th>Sequencing depth</th>
</tr>
</thead>
<tbody>
<tr>
<td>SRR2589044</td>
<td>REL2181A</td>
<td>5,000</td>
<td>Unknown</td>
<td>None</td>
<td>150</td>
<td>60.2</td>
</tr>
<tr>
<td>SRR2584863</td>
<td>REL7179B</td>
<td>15,000</td>
<td>Unknown</td>
<td>None</td>
<td>150</td>
<td>88</td>
</tr>
<tr>
<td>SRR2584866</td>
<td>REL11365</td>
<td>50,000</td>
<td>Cit+</td>
<td>plus</td>
<td>plus</td>
<td>138.3</td>
</tr>
</tbody>
</table>
<p>We want to be able to look at differences in mutation rates between
hypermutable and non-hypermutable strains. We also want to analyze the
sequences to figure out what changes occurred in genomes to make the
strains Cit+. Ultimately, we will use R to answer these questions:</p>
<ul>
<li>How many base pair changes are there between the Cit+ and Cit-
    strains?</li>
<li>What are the base pair changes between strains?</li>
</ul>
<h2 id="how-vcf-files-are-generated">How VCF files are generated<a class="headerlink" href="#how-vcf-files-are-generated" title="Permanent link">&para;</a></h2>
<p>Publicly accessible sequencing files in FASTQ formats can be downloaded
from NCBI SRA. However, FASTQ files contain unaligned sequences of
varying quality, and require clean up and alignment steps before variants
can be called from the reference genome.</p>
<p>There are five steps we must take to transform raw FASTQ files into variant calls (VCF files). At each of the five steps we will be using specialized, non-R based bioinformatics tools:</p>
<figure>
<p><a class="glightbox" href="../figures/variant_calling_workflow.png" data-type="image" data-width="auto" data-height="auto" data-desc-position="bottom"><img alt="image" src="../figures/variant_calling_workflow.png" width="250" /></a>
  </p>
<figcaption>Variant calling workflow</figcaption>
</figure>
<h2 id="how-variant-calls-are-stored-in-vcf-files">How variant calls are stored in VCF files<a class="headerlink" href="#how-variant-calls-are-stored-in-vcf-files" title="Permanent link">&para;</a></h2>
<p>VCF files contain variants that were called against a reference genome.
These files are slightly more complicated than regular tables you can
open using programs like Excel and contain two sections: header and
records.</p>
<p>Below you will see the header (which describes the format), the time and
date the file was created, the version of <code>bcftools</code> that was used, the
command line parameters used, and some additional information:</p>
<div class="admonition quote">
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>##fileformat=VCFv4.2
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>##FILTER=&lt;ID=PASS,Description=&quot;All filters passed&quot;&gt;
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>##bcftoolsVersion=1.8+htslib-1.8
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>##bcftoolsCommand=mpileup -O b -o results/bcf/SRR2584866_raw.bcf -f data/ref_genome/ecoli_rel606.fasta results/bam/SRR2584866.aligned.sorted.bam
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>##reference=file://data/ref_genome/ecoli_rel606.fasta
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>##contig=&lt;ID=CP000819.1,length=4629812&gt;
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a>##ALT=&lt;ID=*,Description=&quot;Represents allele(s) other than observed.&quot;&gt;
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a>##INFO=&lt;ID=INDEL,Number=0,Type=Flag,Description=&quot;Indicates that the variant is an INDEL.&quot;&gt;
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>##INFO=&lt;ID=IDV,Number=1,Type=Integer,Description=&quot;Maximum number of reads supporting an indel&quot;&gt;
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>##INFO=&lt;ID=IMF,Number=1,Type=Float,Description=&quot;Maximum fraction of reads supporting an indel&quot;&gt;
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>##INFO=&lt;ID=DP,Number=1,Type=Integer,Description=&quot;Raw read depth&quot;&gt;
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>##INFO=&lt;ID=VDB,Number=1,Type=Float,Description=&quot;Variant Distance Bias for filtering splice-site artefacts in RNA-seq data (bigger is better)&quot;,Version=
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>##INFO=&lt;ID=RPB,Number=1,Type=Float,Description=&quot;Mann-Whitney U test of Read Position Bias (bigger is better)&quot;&gt;
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>##INFO=&lt;ID=MQB,Number=1,Type=Float,Description=&quot;Mann-Whitney U test of Mapping Quality Bias (bigger is better)&quot;&gt;
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>##INFO=&lt;ID=BQB,Number=1,Type=Float,Description=&quot;Mann-Whitney U test of Base Quality Bias (bigger is better)&quot;&gt;
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>##INFO=&lt;ID=MQSB,Number=1,Type=Float,Description=&quot;Mann-Whitney U test of Mapping Quality vs Strand Bias (bigger is better)&quot;&gt;
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>##INFO=&lt;ID=SGB,Number=1,Type=Float,Description=&quot;Segregation based metric.&quot;&gt;
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>##INFO=&lt;ID=MQ0F,Number=1,Type=Float,Description=&quot;Fraction of MQ0 reads (smaller is better)&quot;&gt;
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>##FORMAT=&lt;ID=PL,Number=G,Type=Integer,Description=&quot;List of Phred-scaled genotype likelihoods&quot;&gt;
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>##FORMAT=&lt;ID=GT,Number=1,Type=String,Description=&quot;Genotype&quot;&gt;
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>##INFO=&lt;ID=ICB,Number=1,Type=Float,Description=&quot;Inbreeding Coefficient Binomial test (bigger is better)&quot;&gt;
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>##INFO=&lt;ID=HOB,Number=1,Type=Float,Description=&quot;Bias in the number of HOMs number (smaller is better)&quot;&gt;
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>##INFO=&lt;ID=AC,Number=A,Type=Integer,Description=&quot;Allele count in genotypes for each ALT allele, in the same order as listed&quot;&gt;
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>##INFO=&lt;ID=AN,Number=1,Type=Integer,Description=&quot;Total number of alleles in called genotypes&quot;&gt;
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>##INFO=&lt;ID=DP4,Number=4,Type=Integer,Description=&quot;Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases&quot;&gt;
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>##INFO=&lt;ID=MQ,Number=1,Type=Integer,Description=&quot;Average mapping quality&quot;&gt;
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>##bcftools_callVersion=1.8+htslib-1.8
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>##bcftools_callCommand=call --ploidy 1 -m -v -o results/bcf/SRR2584866_variants.vcf results/bcf/SRR2584866_raw.bcf; Date=Tue Oct  9 18:48:10 2018
</code></pre></div>
</div>
<p>Followed by information on each of the variations observed:</p>
<div class="admonition quote">
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  results/bam/SRR2584866.aligned.sorted.bam
<a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a>CP000819.1      1521    .       C       T       207     .       DP=9;VDB=0.993024;SGB=-0.662043;MQSB=0.974597;MQ0F=0;AC=1;AN=1;DP4=0,0,4,5;MQ=60
<a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a>CP000819.1      1612    .       A       G       225     .       DP=13;VDB=0.52194;SGB=-0.676189;MQSB=0.950952;MQ0F=0;AC=1;AN=1;DP4=0,0,6,5;MQ=60
<a id="__codelineno-1-4" name="__codelineno-1-4" href="#__codelineno-1-4"></a>CP000819.1      9092    .       A       G       225     .       DP=14;VDB=0.717543;SGB=-0.670168;MQSB=0.916482;MQ0F=0;AC=1;AN=1;DP4=0,0,7,3;MQ=60
<a id="__codelineno-1-5" name="__codelineno-1-5" href="#__codelineno-1-5"></a>CP000819.1      9972    .       T       G       214     .       DP=10;VDB=0.022095;SGB=-0.670168;MQSB=1;MQ0F=0;AC=1;AN=1;DP4=0,0,2,8;MQ=60      GT:PL
<a id="__codelineno-1-6" name="__codelineno-1-6" href="#__codelineno-1-6"></a>CP000819.1      10563   .       G       A       225     .       DP=11;VDB=0.958658;SGB=-0.670168;MQSB=0.952347;MQ0F=0;AC=1;AN=1;DP4=0,0,5,5;MQ=60
<a id="__codelineno-1-7" name="__codelineno-1-7" href="#__codelineno-1-7"></a>CP000819.1      22257   .       C       T       127     .       DP=5;VDB=0.0765947;SGB=-0.590765;MQSB=1;MQ0F=0;AC=1;AN=1;DP4=0,0,2,3;MQ=60      GT:PL
<a id="__codelineno-1-8" name="__codelineno-1-8" href="#__codelineno-1-8"></a>CP000819.1      38971   .       A       G       225     .       DP=14;VDB=0.872139;SGB=-0.680642;MQSB=1;MQ0F=0;AC=1;AN=1;DP4=0,0,4,8;MQ=60      GT:PL
<a id="__codelineno-1-9" name="__codelineno-1-9" href="#__codelineno-1-9"></a>CP000819.1      42306   .       A       G       225     .       DP=15;VDB=0.969686;SGB=-0.686358;MQSB=1;MQ0F=0;AC=1;AN=1;DP4=0,0,5,9;MQ=60      GT:PL
<a id="__codelineno-1-10" name="__codelineno-1-10" href="#__codelineno-1-10"></a>CP000819.1      45277   .       A       G       225     .       DP=15;VDB=0.470998;SGB=-0.680642;MQSB=0.95494;MQ0F=0;AC=1;AN=1;DP4=0,0,7,5;MQ=60
<a id="__codelineno-1-11" name="__codelineno-1-11" href="#__codelineno-1-11"></a>CP000819.1      56613   .       C       G       183     .       DP=12;VDB=0.879703;SGB=-0.676189;MQSB=1;MQ0F=0;AC=1;AN=1;DP4=0,0,8,3;MQ=60      GT:PL
<a id="__codelineno-1-12" name="__codelineno-1-12" href="#__codelineno-1-12"></a>CP000819.1      62118   .       A       G       225     .       DP=19;VDB=0.414981;SGB=-0.691153;MQSB=0.906029;MQ0F=0;AC=1;AN=1;DP4=0,0,8,10;MQ=59
<a id="__codelineno-1-13" name="__codelineno-1-13" href="#__codelineno-1-13"></a>CP000819.1      64042   .       G       A       225     .       DP=18;VDB=0.451328;SGB=-0.689466;MQSB=1;MQ0F=0;AC=1;AN=1;DP4=0,0,7,9;MQ=60      GT:PL
</code></pre></div>
</div>
<p>The first few columns represent the information we have about a
predicted variation.</p>
<table>
<thead>
<tr>
<th>Column</th>
<th>Information</th>
</tr>
</thead>
<tbody>
<tr>
<td>CHROM</td>
<td>Contig location where the variation occurs</td>
</tr>
<tr>
<td>POS</td>
<td>Position within the contig where the variation occurs</td>
</tr>
<tr>
<td>ID</td>
<td>A <code>.</code> until we add annotation information</td>
</tr>
<tr>
<td>REF</td>
<td>Reference genotype (forward strand)</td>
</tr>
<tr>
<td>ALT</td>
<td>Sample genotype (forward strand)</td>
</tr>
<tr>
<td>QUAL</td>
<td>Phred-scaled probability that the observed variant exists at this site (higher is better)</td>
</tr>
<tr>
<td>FILTER</td>
<td>A <code>.</code> if no quality filters have been applied, PASS if a filter is passed, or the name of the filters this variant failed</td>
</tr>
</tbody>
</table>
<p>In an ideal world, the information in the <code>QUAL</code> column would be all we
needed to filter out bad variant calls. However, in reality we need to
filter on multiple other metrics.</p>
<p>The last two columns contain the genotypes and can be tricky to decode.</p>
<table>
<thead>
<tr>
<th>Column</th>
<th>Information</th>
</tr>
</thead>
<tbody>
<tr>
<td>FORMAT</td>
<td>Lists in order the metrics presented int he final column</td>
</tr>
<tr>
<td>results</td>
<td>Lists the values associated with those metrics in order</td>
</tr>
</tbody>
</table>
<p>For our file, the metrics presented are GT:PL:GQ.</p>
<table>
<thead>
<tr>
<th>Metric</th>
<th>Definition</th>
</tr>
</thead>
<tbody>
<tr>
<td>AD, DP</td>
<td>The depth per allele by sample and coverage</td>
</tr>
<tr>
<td>GT</td>
<td>The genotype for the sample at this loci. For a diploid organism, this field indicates the two alleles carried by the sample, encoded by a 0 for the REF allele, 1 for the first ALT allele, 2 for the second ALT allele, etc. A 0/0 means homozygous reference, 0/1 is heterozygous, and 1/1 is homozygous for the alternate allele.</td>
</tr>
<tr>
<td>PL</td>
<td>The likelihoods of the given genotypes.</td>
</tr>
<tr>
<td>GQ</td>
<td>The Phred-scaled confidence for the genotype.</td>
</tr>
</tbody>
</table>
<p>For more information on VCF files visit The Broad Institute's <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035531692-VCF-Variant-Call-Format">VCF
guide</a>.</p>
<div class="admonition book">
<p class="admonition-title">References</p>
<p>Tenaillon O, Barrick JE, Ribeck N, Deatherage DE, Blanchard JL, Dasgupta
A, Wu GC, Wielgoss S, Cruveiller S, Médigue C, Schneider D, Lenski RE.
Tempo and mode of genome evolution in a 50,000-generation experiment
(2016) Nature. 536(7615): 165–170. <br>
<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4988878/">Paper</a>,
<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4988878/#">Supplemental materials</a>, 
<a href="https://trace.ncbi.nlm.nih.gov/Traces/sra/?study=SRP064605">Data on NCBI SRA</a>, 
<a href="https://www.ebi.ac.uk/ena/data/view/PRJNA295606">Data on EMBL-EBI ENA</a></p>
<p>This episode was adapted from the Data Carpentry Genomic lessons:</p>
<ul>
<li><a href="https://datacarpentry.org/organization-genomics/data/">Project organization and management for
    Genomics</a></li>
<li><a href="https://datacarpentry.org/wrangling-genomics/04-variant_calling/index.html">Data wrangling and processing for
    genomics</a></li>
</ul>
</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "..", "features": ["navigation.tracking", "navigation.sections", "navigation.expand", "navigation.top"], "search": "../assets/javascripts/workers/search.d50fe291.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.50899def.min.js"></script>
      
        <script src="../javascripts/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  <script id="init-glightbox">const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});
document$.subscribe(() => { lightbox.reload() });
</script></body>
</html>